File: src/launch_pkg/launch_pkg/boundary_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Point
from time import sleep

class BoundaryPublisher(Node):
    def __init__(self):
        super().__init__('boundary_publisher')
        self.publisher = self.create_publisher(Point, 'boundary_points', 10)
        
        # Timer to publish boundary points
        self.timer = self.create_timer(1.0, self.publish_boundaries)
        
        # Example boundary points for the TurtleSim
        self.boundary_points = [
            (2.0, 2.0),
            (2.0, -2.0),
            (-2.0, -2.0),
            (-2.0, 2.0)
        ]
        self.index = 0

    def publish_boundaries(self):
        if self.index < len(self.boundary_points):
            point = Point()
            point.x = self.boundary_points[self.index][0]
            point.y = self.boundary_points[self.index][1]
            self.publisher.publish(point)
            self.get_logger().info(f"Published boundary point: {point.x}, {point.y}")
            self.index += 1
        else:
            self.get_logger().info("All boundary points published.")
            self.timer.cancel()

def main(args=None):
    rclpy.init(args=args)
    boundary_publisher = BoundaryPublisher()
    try:
        rclpy.spin(boundary_publisher)
    except KeyboardInterrupt:
        boundary_publisher.get_logger().info('Boundary Publisher shutting down...')
    finally:
        boundary_publisher.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/boundry_publisher/src/boundary_publisher.cpp
// src/boundary_publisher_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/point.hpp"
#include <fstream>
#include <sstream>
#include <string>
#include <vector>

class BoundaryPublisher : public rclcpp::Node {
public:
    BoundaryPublisher()
    : Node("boundary_publisher")
    {
        publisher_ = this->create_publisher<geometry_msgs::msg::Point>("boundary_points", 10);
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&BoundaryPublisher::publishNextPoint, this)
        );
        //call path
        loadCoordinates("/home/kemo/Azza_ws/src/boundry_publisher/src/coordinates.csv");
    }

private:
    void loadCoordinates(const std::string& filepath) {
        std::ifstream file(filepath);
        if (!file.is_open()) {
            RCLCPP_ERROR(this->get_logger(), "Failed to open CSV file: %s", filepath.c_str());
            return;
        }

        std::string line;
        while (std::getline(file, line)) {
            std::stringstream ss(line);
            std::string lat, lon;
            if (std::getline(ss, lat, ',') && std::getline(ss, lon, ',')) {
                geometry_msgs::msg::Point point;
                point.x = std::stod(lat);  // latitude
                point.y = std::stod(lon);  // longitude
                point.z = 0.0;
                coordinates_.push_back(point);
            }
        }
        RCLCPP_INFO(this->get_logger(), "Loaded %zu coordinates from CSV", coordinates_.size());
    }

    void publishNextPoint() {
        if (index_ < coordinates_.size()) {
            publisher_->publish(coordinates_[index_]);
            ++index_;
        } else {
            RCLCPP_INFO(this->get_logger(), "All boundary points published.");
            rclcpp::shutdown();
        }
    }

    std::vector<geometry_msgs::msg::Point> coordinates_;
    size_t index_ = 0;
    rclcpp::Publisher<geometry_msgs::msg::Point>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<BoundaryPublisher>());
    rclcpp::shutdown();
    return 0;
}
